src/coreclr/jit/emit.cpp

```diff
@@ -8163,35 +8163,123 @@ CORINFO_FIELD_HANDLE emitter::emitSimd16Const(simd16_t constValue)
+//
+// Arguments:
+//    constValue - constant value
+//    attr       - The EA_SIZE for the constant type
+//
+// Return Value:
+//    A field handle representing the data offset to access the constant.
+//
+// Note:
+// Access to inline data is 'abstracted' by a special type of static member
+// (produced by eeFindJitDataOffs) which the emitter recognizes as being a reference
+// to constant data, not a real static field.
+//
+CORINFO_FIELD_HANDLE emitter::emitSimdConst(simd_t* constValue, emitAttr attr)
 {
-    unsigned cnsSize  = 32;
-    unsigned cnsAlign = cnsSize;
+    unsigned  cnsSize  = EA_SIZE(attr);
+    unsigned  cnsAlign = cnsSize;
+    var_types dataType = (cnsSize >= 8) ? emitComp->getSIMDTypeForSize(cnsSize) : TYP_FLOAT;
 
+#ifdef TARGET_XARCH
     if (emitComp->compCodeOpt() == Compiler::SMALL_CODE)
     {
         cnsAlign = dataSection::MIN_DATA_ALIGN;
     }
+#endif // TARGET_XARCH
 
-    UNATIVE_OFFSET cnum = emitDataConst(&constValue, cnsSize, cnsAlign, TYP_SIMD32);
+    UNATIVE_OFFSET cnum = emitDataConst(constValue, cnsSize, cnsAlign, dataType);
     return emitComp->eeFindJitDataOffs(cnum);
 }
 
-CORINFO_FIELD_HANDLE emitter::emitSimd64Const(simd64_t constValue)
+//------------------------------------------------------------------------
+// emitSimdConstCompressedLoad: Create a simd data section constant,
+//   compressing it if possible, and emit an appropiate instruction
+//   to load or broadcast the constant to a register.
+//
+// Arguments:
+//    constValue - constant value
+//    attr       - The EA_SIZE for the constant type
+//    targetReg  - The target register
+//
+void emitter::emitSimdConstCompressedLoad(simd_t* constValue, emitAttr attr, regNumber targetReg)
 {
-    unsigned cnsSize  = 64;
-    unsigned cnsAlign = cnsSize;
+    assert(EA_SIZE(attr) >= 8);
+    unsigned    cnsSize = EA_SIZE(attr);
+    instruction ins     = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
 
-    if (emitComp->compCodeOpt() == Compiler::SMALL_CODE)
+    for (unsigned size = 4; size < cnsSize; size *= 2)
     {
-        cnsAlign = dataSection::MIN_DATA_ALIGN;
+        // If the constant has zero upper elements/lanes, we can use a smaller load,
+        // because all scalar and vector loads from memory zero the upper.
+
+        simd_t val = {};
+        memcpy(&val, constValue, size);
+
+        if (memcmp(&val, constValue, cnsSize) == 0)
+        {
+            switch (size)
+            {
+                case 4:
+                    ins = INS_movss;
+                    break;
+                case 8:
+                    ins = INS_movsd_simd;
+                    break;
+                default:
+                    ins = INS_movups;
+                    break;
+            }
+
+            cnsSize = size;
+            attr    = EA_ATTR(size);
+            break;
+        }
+
+        if (((size == 8) && (cnsSize == 16) && emitComp->compOpportunisticallyDependsOn(InstructionSet_SSE3)) ||
```

Comment:
Similar here. This "seems" like it can be a check for `64->32->16->8->4->2->1` where it's checking if the two halves are the same, rather than zero.

If they are the same, it can reduce the size and change to the next lower broadcast instruction.

Reply:
Yeah, it's doing that just in reverse. It starts at 4 bytes (there's no point in testing 1 or 2 because relocs have min 4-byte alignment in the emitter anyway) and doubles until the pattern repeats or it reaches the full size.